widget/entry_test.go

```diff
@@ -1616,7 +1616,7 @@ func TestEntry_Select(t *testing.T) {
 		},
 		// Erase the selection and add a newline at selection start
 		"enter": {
-			keys:     []fyne.KeyName{fyne.KeyEnter},
+			keys:     []fyne.KeyName{keyShiftLeftUp, fyne.KeyEnter},
```

Comment:
Sorry if this is a stupid question., but why is the left shift key added here (and below)?

Reply:
Not a stupid question. When I first added the changed to Entry these tests started failing, when I looked into it I found that `setupSelection` was executing `typeKeys(e, keyShiftLeftDown, fyne.KeyRight, fyne.KeyRight, fyne.KeyRight)` and noticed that shift was never getting released. So I added `keyShiftLeftUp` to `setupSelection` and that broke _loads_ of tests and so I moved it to only the tests that failed. It may not be the right approach. @toaster I believe you authored this, could you help us understand this more?

Reply:
I think you realise this, but just in case, we should not be tidying up other tests in new tests.
The source of the error should be corrected.

Reply:
You would have preferred the PR was uploaded and the tests fail when GitHub Actions are run?

Reply:
The source of the error (setupSelection doesn't release the shift key) is in the test, so can you explain what you are expecting me to do?

Reply:
Sorry if it was ambiguous, I would have preferred that the source of the problem was fixed, instead of worked around.
I don't see how that change cannot be included in this PR so that your test does not contain workarounds?

Reply:
As I explained in my first message, I tried to fix setupSelection to release the shift key, but this resulted in more tests breaking - apparently they depended on this behaviour.

Reply:
> can you explain what you are expecting me to do?

When the code that presses shift is done, issue shift up.
Then your test can issue the keys that are required to test the code you are adding.

Reply:
> As I explained in my first message, I tried to fix setupSelection to release the shift key, but this resulted in more tests breaking - apparently they depended on this behaviour.

Yes, I read that, but I still think we should fix the problem not plaster over it.

Reply:
> When the code that presses shift is done

Okay so when widget.entry_test.go setupSelection is done is line 1778

> issue shift up

That breaks:

```
--- FAIL: TestEntry_Select (0.01s)
    --- FAIL: TestEntry_Select/delete_reverse_select_down_with_Shift_still_hold (0.00s)
        entry_test.go:1789: 
            	Error Trace:	entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "ng\nTe"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1 @@
            	            	-ng
            	            	-Te
            	            	+
            	Test:       	TestEntry_Select/delete_reverse_select_down_with_Shift_still_hold
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,29\" size=\"20x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,50\" size=\"18x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -3,4 +3,2 @@
            	            	 		<widget pos="10,10" size="120x100" type="*widget.Entry">
            	            	-			<rectangle fillColor="focus" pos="24,29" size="20x21"/>
            	            	-			<rectangle fillColor="focus" pos="7,50" size="18x21"/>
            	            	 			<rectangle fillColor="focus" pos="0,96" size="120x4"/>
            	Test:       	TestEntry_Select/delete_reverse_select_down_with_Shift_still_hold
    --- FAIL: TestEntry_Select/backspace_reverse_select_down_with_Shift_still_hold (0.00s)
        entry_test.go:1789: 
            	Error Trace:	entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "ng\nTe"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1 @@
            	            	-ng
            	            	-Te
            	            	+
            	Test:       	TestEntry_Select/backspace_reverse_select_down_with_Shift_still_hold
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,29\" size=\"20x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,50\" size=\"18x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -3,4 +3,2 @@
            	            	 		<widget pos="10,10" size="120x100" type="*widget.Entry">
            	            	-			<rectangle fillColor="focus" pos="24,29" size="20x21"/>
            	            	-			<rectangle fillColor="focus" pos="7,50" size="18x21"/>
            	            	 			<rectangle fillColor="focus" pos="0,96" size="120x4"/>
            	Test:       	TestEntry_Select/backspace_reverse_select_down_with_Shift_still_hold
    --- FAIL: TestEntry_Select/delete_reverse_multi-line (0.00s)
        entry_test.go:1788: 
            	Error Trace:	entry_test.go:1788
            	Error:      	Not equal: 
            	            	expected: "Testing\nTestisting"
            	            	actual  : "Testing\nTesting\nTeting"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	 Testing
            	            	-Testisting
            	            	+Testing
            	            	+Teting
            	Test:       	TestEntry_Select/delete_reverse_multi-line
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Testisting</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\"></text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"41,29\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Teting</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -6,6 +6,6 @@
            	            	 				<text pos="4,4" size="104x21">Testing</text>
            	            	-				<text pos="4,25" size="104x21">Testisting</text>
            	            	-				<text pos="4,46" size="104x21"></text>
            	            	+				<text pos="4,25" size="104x21">Testing</text>
            	            	+				<text pos="4,46" size="104x21">Teting</text>
            	            	 			</widget>
            	            	-			<rectangle fillColor="focus" pos="41,29" size="2x21"/>
            	            	+			<rectangle fillColor="focus" pos="24,50" size="2x21"/>
            	            	 		</widget>
            	Test:       	TestEntry_Select/delete_reverse_multi-line
    --- FAIL: TestEntry_Select/delete_reverse_select_up_with_Shift_still_hold (0.00s)
        entry_test.go:1789: 
            	Error Trace:	entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "sting\nTe"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1 @@
            	            	-sting
            	            	-Te
            	            	+
            	Test:       	TestEntry_Select/delete_reverse_select_up_with_Shift_still_hold
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"37x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,29\" size=\"18x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -3,4 +3,2 @@
            	            	 		<widget pos="10,10" size="120x100" type="*widget.Entry">
            	            	-			<rectangle fillColor="focus" pos="24,8" size="37x21"/>
            	            	-			<rectangle fillColor="focus" pos="7,29" size="18x21"/>
            	            	 			<rectangle fillColor="focus" pos="0,96" size="120x4"/>
            	Test:       	TestEntry_Select/delete_reverse_select_up_with_Shift_still_hold
    --- FAIL: TestEntry_Select/backspace_multi-line (0.00s)
        entry_test.go:1788: 
            	Error Trace:	entry_test.go:1788
            	Error:      	Not equal: 
            	            	expected: "Testing\nTeng"
            	            	actual  : "Testing\nTesting\nTestng"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	 Testing
            	            	-Teng
            	            	+Testing
            	            	+Testng
            	Test:       	TestEntry_Select/backspace_multi-line
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\"></text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,29\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testng</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"37,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -6,6 +6,6 @@
            	            	 				<text pos="4,4" size="104x21">Testing</text>
            	            	-				<text pos="4,25" size="104x21">Teng</text>
            	            	-				<text pos="4,46" size="104x21"></text>
            	            	+				<text pos="4,25" size="104x21">Testing</text>
            	            	+				<text pos="4,46" size="104x21">Testng</text>
            	            	 			</widget>
            	            	-			<rectangle fillColor="focus" pos="24,29" size="2x21"/>
            	            	+			<rectangle fillColor="focus" pos="37,50" size="2x21"/>
            	            	 		</widget>
            	Test:       	TestEntry_Select/backspace_multi-line
    --- FAIL: TestEntry_Select/backspace_reverse_multi-line (0.00s)
        entry_test.go:1788: 
            	Error Trace:	entry_test.go:1788
            	Error:      	Not equal: 
            	            	expected: "Testing\nTestisting"
            	            	actual  : "Testing\nTesting\nTsting"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	 Testing
            	            	-Testisting
            	            	+Testing
            	            	+Tsting
            	Test:       	TestEntry_Select/backspace_reverse_multi-line
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Testisting</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\"></text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"41,29\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Tsting</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"16,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -6,6 +6,6 @@
            	            	 				<text pos="4,4" size="104x21">Testing</text>
            	            	-				<text pos="4,25" size="104x21">Testisting</text>
            	            	-				<text pos="4,46" size="104x21"></text>
            	            	+				<text pos="4,25" size="104x21">Testing</text>
            	            	+				<text pos="4,46" size="104x21">Tsting</text>
            	            	 			</widget>
            	            	-			<rectangle fillColor="focus" pos="41,29" size="2x21"/>
            	            	+			<rectangle fillColor="focus" pos="16,50" size="2x21"/>
            	            	 		</widget>
            	Test:       	TestEntry_Select/backspace_reverse_multi-line
    --- FAIL: TestEntry_Select/backspace_select_down_with_Shift_still_hold (0.00s)
        entry_test.go:1789: 
            	Error Trace:	entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "ng\nTe"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1 @@
            	            	-ng
            	            	-Te
            	            	+
            	Test:       	TestEntry_Select/backspace_select_down_with_Shift_still_hold
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,29\" size=\"20x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,50\" size=\"18x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -3,4 +3,2 @@
            	            	 		<widget pos="10,10" size="120x100" type="*widget.Entry">
            	            	-			<rectangle fillColor="focus" pos="24,29" size="20x21"/>
            	            	-			<rectangle fillColor="focus" pos="7,50" size="18x21"/>
            	            	 			<rectangle fillColor="focus" pos="0,96" size="120x4"/>
            	Test:       	TestEntry_Select/backspace_select_down_with_Shift_still_hold
    --- FAIL: TestEntry_Select/delete_select_down_with_Shift_still_hold (0.00s)
        entry_test.go:1789: 
            	Error Trace:	entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "ng\nTe"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1 @@
            	            	-ng
            	            	-Te
            	            	+
            	Test:       	TestEntry_Select/delete_select_down_with_Shift_still_hold
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,29\" size=\"20x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,50\" size=\"18x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -3,4 +3,2 @@
            	            	 		<widget pos="10,10" size="120x100" type="*widget.Entry">
            	            	-			<rectangle fillColor="focus" pos="24,29" size="20x21"/>
            	            	-			<rectangle fillColor="focus" pos="7,50" size="18x21"/>
            	            	 			<rectangle fillColor="focus" pos="0,96" size="120x4"/>
            	Test:       	TestEntry_Select/delete_select_down_with_Shift_still_hold
    --- FAIL: TestEntry_Select/delete_multi-line (0.00s)
        entry_test.go:1788: 
            	Error Trace:	entry_test.go:1788
            	Error:      	Not equal: 
            	            	expected: "Testing\nTeng"
            	            	actual  : "Testing\nTesting\nTestig"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	 Testing
            	            	-Teng
            	            	+Testing
            	            	+Testig
            	Test:       	TestEntry_Select/delete_multi-line
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\"></text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,29\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testig</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"41,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -6,6 +6,6 @@
            	            	 				<text pos="4,4" size="104x21">Testing</text>
            	            	-				<text pos="4,25" size="104x21">Teng</text>
            	            	-				<text pos="4,46" size="104x21"></text>
            	            	+				<text pos="4,25" size="104x21">Testing</text>
            	            	+				<text pos="4,46" size="104x21">Testig</text>
            	            	 			</widget>
            	            	-			<rectangle fillColor="focus" pos="24,29" size="2x21"/>
            	            	+			<rectangle fillColor="focus" pos="41,50" size="2x21"/>
            	            	 		</widget>
            	Test:       	TestEntry_Select/delete_multi-line
    --- FAIL: TestEntry_Select/backspace_select_up_with_Shift_still_hold (0.00s)
        entry_test.go:1789: 
            	Error Trace:	entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "sting\nTe"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1 @@
            	            	-sting
            	            	-Te
            	            	+
            	Test:       	TestEntry_Select/backspace_select_up_with_Shift_still_hold
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"37x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,29\" size=\"18x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -3,4 +3,2 @@
            	            	 		<widget pos="10,10" size="120x100" type="*widget.Entry">
            	            	-			<rectangle fillColor="focus" pos="24,8" size="37x21"/>
            	            	-			<rectangle fillColor="focus" pos="7,29" size="18x21"/>
            	            	 			<rectangle fillColor="focus" pos="0,96" size="120x4"/>
            	Test:       	TestEntry_Select/backspace_select_up_with_Shift_still_hold
    --- FAIL: TestEntry_Select/backspace_reverse_select_up_with_Shift_still_hold (0.00s)
        entry_test.go:1789: 
            	Error Trace:	entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "sting\nTe"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1 @@
            	            	-sting
            	            	-Te
            	            	+
            	Test:       	TestEntry_Select/backspace_reverse_select_up_with_Shift_still_hold
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"37x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,29\" size=\"18x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -3,4 +3,2 @@
            	            	 		<widget pos="10,10" size="120x100" type="*widget.Entry">
            	            	-			<rectangle fillColor="focus" pos="24,8" size="37x21"/>
            	            	-			<rectangle fillColor="focus" pos="7,29" size="18x21"/>
            	            	 			<rectangle fillColor="focus" pos="0,96" size="120x4"/>
            	Test:       	TestEntry_Select/backspace_reverse_select_up_with_Shift_still_hold
    --- FAIL: TestEntry_Select/delete_select_up_with_Shift_still_hold (0.00s)
        entry_test.go:1789: 
            	Error Trace:	entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "sting\nTe"
            	            	actual  : ""
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1 @@
            	            	-sting
            	            	-Te
            	            	+
            	Test:       	TestEntry_Select/delete_select_up_with_Shift_still_hold
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1790
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"37x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,29\" size=\"18x21\"/>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,8\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -3,4 +3,2 @@
            	            	 		<widget pos="10,10" size="120x100" type="*widget.Entry">
            	            	-			<rectangle fillColor="focus" pos="24,8" size="37x21"/>
            	            	-			<rectangle fillColor="focus" pos="7,29" size="18x21"/>
            	            	 			<rectangle fillColor="focus" pos="0,96" size="120x4"/>
            	Test:       	TestEntry_Select/delete_select_up_with_Shift_still_hold
```

Reply:
If I remove the additions that this thread is referring to:
```
diff --git a/widget/entry_test.go b/widget/entry_test.go
index 48235a2b..fb784d73 100644
--- a/widget/entry_test.go
+++ b/widget/entry_test.go
@@ -1616,7 +1616,7 @@ func TestEntry_Select(t *testing.T) {
                },
                // Erase the selection and add a newline at selection start
                "enter": {
-                       keys:     []fyne.KeyName{keyShiftLeftUp, fyne.KeyEnter},
+                       keys:     []fyne.KeyName{fyne.KeyEnter},
                        wantText: "Testing\nTe\nng\nTesting",
                        wantMarkup: `
        <canvas padded size="150x200">
@@ -1636,7 +1636,7 @@ func TestEntry_Select(t *testing.T) {
                        `,
                },
                "enter reverse": {
-                       keys:         []fyne.KeyName{keyShiftLeftUp, fyne.KeyEnter},
+                       keys:         []fyne.KeyName{fyne.KeyEnter},
                        setupReverse: true,
                        wantText:     "Testing\nTe\nng\nTesting",
                        wantMarkup: `
```
It breaks:
```
--- FAIL: TestEntry_Select (0.01s)
    --- FAIL: TestEntry_Select/enter (0.00s)
        entry_test.go:1787: 
            	Error Trace:	entry_test.go:1787
            	Error:      	Not equal: 
            	            	expected: "Testing\nTe\nng\nTesting"
            	            	actual  : "Testing\nTeng\nTesting"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,4 +1,3 @@
            	            	 Testing
            	            	-Te
            	            	-ng
            	            	+Teng
            	            	 Testing
            	Test:       	TestEntry_Select/enter
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Te</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">ng</text>\n\t\t\t\t<text pos=\"4,67\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,29\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -6,7 +6,6 @@
            	            	 				<text pos="4,4" size="104x21">Testing</text>
            	            	-				<text pos="4,25" size="104x21">Te</text>
            	            	-				<text pos="4,46" size="104x21">ng</text>
            	            	-				<text pos="4,67" size="104x21">Testing</text>
            	            	+				<text pos="4,25" size="104x21">Teng</text>
            	            	+				<text pos="4,46" size="104x21">Testing</text>
            	            	 			</widget>
            	            	-			<rectangle fillColor="focus" pos="7,50" size="2x21"/>
            	            	+			<rectangle fillColor="focus" pos="24,29" size="2x21"/>
            	            	 		</widget>
            	Test:       	TestEntry_Select/enter
    --- FAIL: TestEntry_Select/enter_reverse (0.00s)
        entry_test.go:1787: 
            	Error Trace:	entry_test.go:1787
            	Error:      	Not equal: 
            	            	expected: "Testing\nTe\nng\nTesting"
            	            	actual  : "Testing\nTeng\nTesting"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,4 +1,3 @@
            	            	 Testing
            	            	-Te
            	            	-ng
            	            	+Teng
            	            	 Testing
            	Test:       	TestEntry_Select/enter_reverse
        test.go:48: 
            	Error Trace:	test.go:48
            	            				entry_test.go:1789
            	Error:      	Not equal: 
            	            	expected: "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Te</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">ng</text>\n\t\t\t\t<text pos=\"4,67\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"7,50\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	actual  : "<canvas padded size=\"150x200\">\n\t<content>\n\t\t<widget pos=\"10,10\" size=\"120x100\" type=\"*widget.Entry\">\n\t\t\t<rectangle fillColor=\"focus\" pos=\"0,96\" size=\"120x4\"/>\n\t\t\t<widget pos=\"4,4\" size=\"112x92\" type=\"*widget.textProvider\">\n\t\t\t\t<text pos=\"4,4\" size=\"104x21\">Testing</text>\n\t\t\t\t<text pos=\"4,25\" size=\"104x21\">Teng</text>\n\t\t\t\t<text pos=\"4,46\" size=\"104x21\">Testing</text>\n\t\t\t</widget>\n\t\t\t<rectangle fillColor=\"focus\" pos=\"24,29\" size=\"2x21\"/>\n\t\t</widget>\n\t</content>\n</canvas>\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -6,7 +6,6 @@
            	            	 				<text pos="4,4" size="104x21">Testing</text>
            	            	-				<text pos="4,25" size="104x21">Te</text>
            	            	-				<text pos="4,46" size="104x21">ng</text>
            	            	-				<text pos="4,67" size="104x21">Testing</text>
            	            	+				<text pos="4,25" size="104x21">Teng</text>
            	            	+				<text pos="4,46" size="104x21">Testing</text>
            	            	 			</widget>
            	            	-			<rectangle fillColor="focus" pos="7,50" size="2x21"/>
            	            	+			<rectangle fillColor="focus" pos="24,29" size="2x21"/>
            	            	 		</widget>
            	Test:       	TestEntry_Select/enter_reverse
```

Reply:
I guess if it is not possible to figure out then hopefully the hint from toaster that you requested will help?

Reply:
I had a quick look to see if I could figure out anything obvious.
It seems like the strange behaviour of leaving the key down is for the test in `TestEntry_SelectHomeEnd` as all the other usages I could see immediately put the key back up before running their tests.
Instead of continuing this anti-pattern it seems like a key-up at the end of setup then another key-down for that one special case would lead to more readable code.
That is not a complete solution, it seems, so there is more going on in there - but that feels like the right approach on the face of it.

Reply:
I think I cracked it.

After removing all the extra keyShiftLeftUp, you could add 
```
			typeKeys(entry, keyShiftLeftDown)
```
around likes 650, 1784 and 1881, having added the up at the end of `setupSelection` the tests all pass again for me

Reply:
Both approaches require test changes and this was the smaller of the two, tho I agree that the right approach is discontinue the ant-pattern and to that end I've created #1730, once it is merged this PR can pull it in so the only test changes are indeed for the feature being added.